### PR TITLE
Change abstract interface

### LINKED ISSUE #41: Make an Optic interface a little more abstract (hide indexed implementation details)
Problem: Talk.pdf says "This just works, because our Lens type isn’t polymorphic"

But it's not true. The true 'Lens' type in current @optics@ is

    forall i. Lens i s t a b

i.e. we __have__ leave it free. Not good.
(We can set it to () or Void, to put into container,
 but then we cannot compose with those optics)

Also the double indices __leak__ the implementation details. Not good either.

Solition: We can cure this, with some additional hackery.
And we can make CheckIndices less hacky, as it would only match on type level list!
I only comment the parts which are changed.

The idea is to use 

    newtype Optic (k :: OptikKind) (is :: [Type) s t a b
        Optic (forall p i. Constraints k p => p i a b -> p (Curry is i) s t)

abstract interface. I.e. move the first (of double indices) to be inside the
Optic.

Note: The implementation is still a compromise.
- A_Setter has always empty index list
- An_IxSettter has always non empty index list
But we are forced to it this way so Constraints reduce properly.  i.e. we can
`sub` Setter into IxSetter, but it would be unusable (as it would have empty
index list.

We can use CheckIndices (and CheckNonIndices if we like) to get better errors.
I think latter is not needed as library doesn't give means to construct
non-indexed optic with non-empty index-list.

More elegantly we'd only have `A_Setter` and differentiate
between indexed and non-indexed variants based on index-list.
But then the type-checker will fight back.

- Making Constraints (k :: OpticKind) (is :: [*]) p breaks `Is` implementation:
  in implies `is` is rigid, so type-family cannot match on it (or we need more
  proofs)

- Using `Constraints An_Traversal p = IxTraversing p`
  but then e.g. `over` would need to be implemented using `IxFunArrow`,
  e.g. passing some dummy index value. Not good.

The full gist is in https://gist.github.com/phadej/0a1b1aa95f102a78d4e26b82f45da903

@arybczak What do you think about this?

### PR DESCRIPTION
**Pull request by phadej**
Resolves #41

I temporarily moved `ixcontramap` to `Traversing` (from `Profunctor` where it belongs).

### CODE REVIEW

File: optics-core/src/Optics/Internal/Optic/TypeLevel.hs [hunk 1]
```diff
@@ -0,0 +1,73 @@
+{-# LANGUAGE TypeApplications #-}
```
Comment (arybczak) [inline optics-core/src/Optics/Internal/Optic/TypeLevel.hs]: Can we keep 2 spaces for indentation everywhere?

File: optics-sop/src/Generics/SOP/Optics.hs [hunk 2]
```diff
@@ -21,49 +21,49 @@ import Generics.SOP
 import Optics hiding (to)
 
 -- | Iso between a generic type and its representation.
-rep :: (Generic a, Generic b) => Iso i a b (Rep a) (Rep b)
+rep :: (Generic a, Generic b) => Iso a b (Rep a) (Rep b)
 rep = iso from to
 
--- | Iso induced by the 'SOP' newtype.
-sop :: Iso i (SOP f xss) (SOP g yss) (NS (NP f) xss) (NS (NP g) yss)
+-- | Isonduced by the 'SOP' newtype.
```
Comment (arybczak) [inline optics-sop/src/Generics/SOP/Optics.hs]: It seems you search-and-replaced a bit too much here ;) Same below.

Comment (phadej) [reply optics-sop/src/Generics/SOP/Optics.hs]: Thanks, fixed

**Comment by arybczak** [comment]
Latest commit changes functions that expect a regular optic to functions that take any optic, ie. both regular and indexed ones, silently ignoring the indices in the latter case. Do we really want it? So far we agreed for explicit conversion only.

**Comment by phadej** [comment]
We agreed with @kosmikus and @adamgundry, that this is ok. Hopefully there weren't any miscommunication.

It's easy to change though with `CheckNonIndexed` class.

**Comment by adamgundry** [comment]
Yes, we have a design choice here, but our most recent thinking is that we should permit using an indexed optic as a regular one, rather than mandating an explicit conversion. This fits with the perspective that indices are somewhat orthogonal to the optic kind: an indexed setter is "just" a setter that can provide indices, but they don't necessarily have to be used. Thus we allow silent downgrading much as we allow a lens to be silently used as a setter, even though it is capable of more.

There doesn't seem to be much harm to the extra flexibility; it allows things like `toListOf ifolded` where `toListOf folded` would do, which might be undesirable in some cases, but in others it is nice not to require separate names for the indexed and non-indexed versions.

@arybczak are there any major downsides of this approach that we are missing?

**Comment by arybczak** [comment]
Alright, I just wanted to make sure we're on the same page.

The only thing I can think about is the fact that indexed versions with unused indices are slower than regular ones. But perhaps we can mitigate that by coming with something similar (yet simpler) to `Conjoined` in the future.

**Comment by phadej** [comment]
Rebased on current `master`.

**Pull request merged**